CDAP-18010: Run artifact inspection in isolation.

wyzhang opened the pull request:
Part 1: Add an endpoint in task worker to allow submit tasks via query params.

Why:
Artifact inspection takes places right after upload completes, it resides
in a temp dir in appfab. To run artifact inspection in worker pod, appfab
needs to send file over. Current rest endpoint "/run" use request body
to store task specification, which makes it challenging to send file
as well. Thus introducing a new endpoint to specify task specifiction
via query parameter, then file content can be streamed over via
request body.

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java:
```diff
@@ -87,6 +100,79 @@ public void run(FullHttpRequest request, HttpResponder responder) {
+            RunnableTaskRequest runnableTaskRequest = new RunnableTaskRequest(className, param);
+            byte[] response = runnableTaskLauncher.launchRunnableTask(runnableTaskRequest, dest.toURI());
+            responder.sendByteArray(HttpResponseStatus.OK,
+                                    response,
```
There is a chance that worker pod stops before finishing to send the response. This can be addressed in a followup PR.

masoud-io commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java:
```diff
@@ -87,6 +100,79 @@ public void run(FullHttpRequest request, HttpResponder responder) {
+   * @return {@link BodyConsumer}
+   */
+  @POST
+  @Path("/task")
```
nit: /run and /task might be a bit confusing. I though don't have a better suggestion either.

masoud-io commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/TaskWorkerServiceTest.java:
```diff
@@ -184,6 +195,37 @@ public void testConcurrentRequests() throws Exception {
+    Assert.assertEquals(content, response.getResponseBodyAsString());
+
+    waitForTaskWorkerToFinish(taskWorkerService);
+    Assert.assertTrue(taskWorkerService.state() == Service.State.TERMINATED);
```
nit: also assert that tmp file gets deleted.

wyzhang committed:
CDAP-18010: Run artifact inspection in isolation.

Part 1: Add an endpoint in task worker to allow submit tasks via query params.
Part 2: Allow new endpoint to allow uploading a file which is made
        available for task runnable via RunnableTaskContext

Why:
Artifact inspection takes places right after upload completes, it resides
in a temp dir in appfab. To run artifact inspection in worker pod, appfab
needs to send file over. Current rest endpoint "/run" use request body
to store task specification, which makes it challenging to send file
as well. Thus introducing a new endpoint to specify task specifiction
via query parameter, then file content can be streamed over via
request body.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java:
```diff
@@ -96,6 +111,71 @@ public void run(FullHttpRequest request, HttpResponder responder) {
+        @Override
+        protected void onError(Throwable cause) {
+          tmpFile.delete();
+          LOG.info("Failed to download file to run task", className);
```
Missing placeholder `{}` for the classname

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java:
```diff
@@ -96,6 +111,71 @@ public void run(FullHttpRequest request, HttpResponder responder) {
+        }
+      };
+    } catch (IOException e) {
+      LOG.error("Failed to download file to run task", className);
```
Missing placeholder `{}` for the classname
